[DA][8/n][user-code] Add simplified string-only `cursor` property

OwenKephart opened the pull request:
## Summary & Motivation

Technically, the Declarative Automation cursoring system can accept and store any serializable value in its "extra_state". This is useful as a performance optimization (it is common for us to want to keep track of serialized AssetSubsets between ticks, so not needing to do a double serialization / deserialization process is pretty valuable), but this overly-broad construct is likely more confusing than anything if we expose it as a public API.

Instead, expose a simplified, string-only cursor interface on the context. This mirrors the existing sensor APIs more closely.

## How I Tested These Changes

OwenKephart commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23596?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#23596** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23596?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#23593** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23593?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#23585** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23585?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#23584** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23584?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#23583** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23583?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#23581** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23581?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#23535** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23535?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>: 1 other dependent PR ([#23536](https://github.com/dagster-io/dagster/pull/23536) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23536?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>)
* **#23534** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23534?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @OwenKephart and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

schrockn commented on python_modules/dagster/dagster/_core/definitions/declarative_automation/automation_condition.py:
```diff
@@ -448,7 +450,12 @@ def __init__(
         self._subsets_with_metadata = check.opt_sequence_param(
             subsets_with_metadata, "subsets_with_metadata", AssetSubsetWithMetadata
         )
-        self._extra_state = extra_state
+        check.param_invariant(
+            not (cursor and extra_state),
+            "extra_state",
+            "Cannot provide both cursor and extra_state.",
+        )
+        self._extra_state = check.opt_str_param(cursor, "cursor") or extra_state
```
<!--__GRAPHITE_HTML_TAG_START__--><i class='graphite__hidden'>[Re: lines +432 to +458]<br/><br/></i>

<!--__GRAPHITE_HTML_TAG_END__-->I think it's pretty non-obvious as to why these should be mutually exclusive given the names etc<!--__GRAPHITE_HTML_TAG_START__--><span class='graphite__hidden'><br/><br/>See this comment inline on <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/23596?utm_source=unchanged-line-comment">Graphite</a>.</span>

<!--__GRAPHITE_HTML_TAG_END__-->

schrockn reviewed (changes requested):
I think this is quite confusing to canonicalize at the layer you are doing it in. If there is only a cursor then I think it should only take a cursor in the constructor.

OwenKephart commented:
@schrockn I've updated the PR to treat `extra_state` (and `subsets_with_metadata`) as a "hidden_param", and added some explanatory comments.

The basic issue at play here is that we have some built-in conditions which need to store complex, but serdes-able objects (namely `AssetSubsets`).

It is possible for those to be serialized into strings and jammed into a cursor, but this would be a pretty significant drag on performance and overall cursor size (as we'd end up double-serializing that cursor, because it exists in the context of a larger structured object that also gets serialized). So this exists as a higher-performance back door for builtin conditions to take advantage of.

However, we want to present a simpler, string-only interface to the user, which is what this PR is doing.

schrockn commented on python_modules/dagster/dagster/_core/definitions/declarative_automation/automation_condition.py:
```diff
@@ -388,10 +388,20 @@ def __init__(
         )
-        self._extra_state = extra_state
+
+        # hidden_param which should only be set by builtin conditions which require high performance
+        # in their serdes layer
+        extra_state = kwargs.get("extra_state")
+        check.param_invariant(
+            not (cursor and extra_state),
+            "extra_state",
+            "Cannot provide both cursor and extra_state.",
+        )
+        self._extra_state = check.opt_str_param(cursor, "cursor") or extra_state
```
Well I think we should offer the same interface to users as well.

What if we instead typed `ExtraState` as `StructuredCursor` then typed

```
cursor: Union[str, StructuredCursor]
```

I think that would make intent more clear and avoid an invalid state that is only caught at runtime.

OwenKephart replied:
@schrockn The issue with that is that the read side gets more complicated as the user (who almost exclusively will just be writing a string value to the cursor) will now have to handle a type that's potentially not a string when they access `context.cursor`.

My feeling is that a bare string cursor has been sufficient for sensors, and so it seems likely that it will be sufficient for AutomationConditions as well. I updated the name of the hidden parameter to `structured_cursor` to make it more obvious that these parameters intersect with each other. My feeling is that if in the future we see demand for higher-performance cursors in user code, we can un-hide this parameter.

schrockn reviewed (changes requested):
inline comment

schrockn commented on python_modules/dagster/dagster/_core/definitions/declarative_automation/automation_context.py:
```diff
@@ -135,14 +138,19 @@ def node_cursor(self) -> Optional[AutomationConditionNodeCursor]:
+    @property
+    def cursor(self) -> Optional[str]:
+        """The cursor value returned on the previous evaluation for this condition, if any."""
+        return self._node_cursor.get_extra_state(as_type=str) if self._node_cursor else None
```
let's rename `get_extra_state` `get_structured_cursor`

schrockn reviewed (approved):
Cool. Please do further renames of extra_state to structured_cursor as appropriate as I think that is much more clear!

OwenKephart commented:
### Merge activity

* **Aug 22, 5:31 PM EDT**: @OwenKephart started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/23596).
